### DISCUSSION TITLE
Zero UV partials for custom BSDF

### DISCUSSION THREAD

**Discussion by msqrt** [mitsuba-renderer/mitsuba3#980] [Q&A] [answered]
Hi! I'm trying to implement a custom SVBRDF model with PyTorch -- everything else seems to be working, but the `duv_dx` and `duv_dy` fields of the surface interaction are always all zeros. I've added the `needs_differentials` function that returns `True`, but it never gets called. This seems to be caused by [this code](https://github.com/mitsuba-renderer/mitsuba3/blob/3549783b37a9d831997f9c39a2a8d5affcbf2c50/include/mitsuba/render/bsdf.h#L642). `has_uv_partials` for the surface interaction seems to be true every time, but I'm not sure what the `dr::is_diff_v<Float>` is, and the comment makes me worried that this is just not supported at all.

Is there something I should configure differently to make this work? Or should I compile Mitsuba myself and try to fix the issue in the linked code?

**Comment by njroussel** [comment, marked as answer]
Hi @msqrt 

Indeed, it seems that this path of code is only enabled in non-AD variants. I wonder why...

You can either re-compile the project yourself and remove that `constexpr ... is_diff_v<Float`  condition. Another solution would be to explicitly call `si.compute_uv_partials` right before calling `si.bsdf(ray)` in your code.

**Reply by msqrt** [reply]
Alright, thanks! I changed that and compiled a custom version and it works now -- I ended up having to change the path tracer as well, since it doesn't propagate the ray differentials even for the first bounce.
